### PR TITLE
U423/t611 join in game

### PR DESCRIPTION
**Pull request by Discojanne**

### CODE REVIEW

File: Sail/src/Sail/entities/systems/network/HostSendToSpectatorSystem.cpp [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+	size_t nrOfEvents = 0;
+	bool candleHeld = true;
+
+	// Find how many players are alive so that the receiver knows how many messages to receive
+	for (auto e : entities) {
+		if (e->getComponent<NetworkSenderComponent>()->m_entityType == Netcode::EntityType::PLAYER_ENTITY) {
+			nrOfEvents++;
+
+			for (auto c : e->getChildEntities()) {
+				if (c->hasComponent<CandleComponent>() && !c->getComponent<CandleComponent>()->isCarried) {
```
Comment (tofb15) [inline Sail/src/Sail/entities/systems/network/HostSendToSpectatorSystem.cpp]: CandleComponent is not a registered component in the constructor.

registerComponent\<CandleComponent\>(false, true, false);

File: Sail/src/Sail/entities/systems/network/HostSendToSpectatorSystem.cpp [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+	ar(size_t{ 0 });
+
+
+
+	// TODO: if a player's candle isn't held also send an event for that
+	size_t nrOfEvents = 0;
+	bool candleHeld = true;
+
+	// Find how many players are alive so that the receiver knows how many messages to receive
+	for (auto e : entities) {
```
Comment (tofb15) [inline Sail/src/Sail/entities/systems/network/HostSendToSpectatorSystem.cpp]: How many entities are looped here? Is it possible to register more components as required to reduse the number of entities in this system without losing the entities you are looking for?

File: Sail/src/Sail/entities/systems/network/HostSendToSpectatorSystem.cpp [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+					nrOfEvents++;
+				}
+			}
+		}
+	}
+
+	ar(nrOfEvents);
+
+	// Loop through the entities and send messages the necessary information needed to create the entities for the spectator
+	for (auto e : entities) {
```
Comment (tofb15) [inline Sail/src/Sail/entities/systems/network/HostSendToSpectatorSystem.cpp]: How many entities are looped here? Is it possible to register more components as required to reduse the number of entities in this system without losing the entities you are looking for?

File: Sail/src/Sail/entities/systems/network/HostSendToSpectatorSystem.cpp [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+	m_playerID = playerID;
+}
+
+
+// The messages this function creates must match the format of messages created by and received by NetworkSenderSystem and NetworkReceiverSystem
+void HostSendToSpectatorSystem::sendEntityCreationPackage(TCP_CONNECTION_ID TCPidPlayer) const {
+	std::ostringstream dataString(std::ios::binary);
+	Netcode::OutArchive ar(dataString);
+
+	// -+-+-+-+-+-+-+-+ Per-frame sends to per-frame receives via components -+-+-+-+-+-+-+-+
```
Comment (tofb15) [inline Sail/src/Sail/entities/systems/network/HostSendToSpectatorSystem.cpp]: is this comment relevant here?

"// -+-+-+-+-+-+-+-+ Per-frame sends to per-frame receives via components -+-+-+-+-+-+-+-+ "

File: Sail/src/Sail/entities/systems/network/HostSendToSpectatorSystem.cpp [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+			Netcode::ComponentID candleID, gunID;
+			// Find the component ID of the player's candle
+			for (auto c : e->getChildEntities()) {
+				if (c->hasComponent<CandleComponent>()) {
+					candleID = c->getComponent<NetworkSenderComponent>()->m_id;
+				}
+				else {
+					if (e->hasComponent<GunComponent>()) {
+						gunID = c->getComponent<NetworkSenderComponent>()->m_id;
+					}
```
Comment (tofb15) [inline Sail/src/Sail/entities/systems/network/HostSendToSpectatorSystem.cpp]: Should this be here in the loop?

File: Sail/src/Sail/entities/systems/network/HostSendToSpectatorSystem.cpp [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+				}
+
+
+			}
+			if (e->hasComponent<GunComponent>()) {
+				gunID = e->getComponent<NetworkSenderComponent>()->m_id;
+			}
```
Comment (tofb15) [inline Sail/src/Sail/entities/systems/network/HostSendToSpectatorSystem.cpp]: Almost the same code as on row [78-80]

File: SPLASH/src/game/states/GameState.cpp [hunk 7]
```diff
@@ -144,14 +147,29 @@ GameState::GameState(StateStack& stack)
 	createLevel(shader, boundingBoxModel);
 
 	// Player creation
+	if (m_app->getStateStorage().getLobbyToGameData()->enterAsSpectator) {
+		//m_player = EntityFactory::CreateMySpectator(playerID).get();
```
Comment (tofb15) [inline SPLASH/src/game/states/GameState.cpp]: Remove out comented code

**Comment by DanielFredriksson** [comment]
Closed as it is under development

**Pull request closed without merging**